### ISSUE TITLE
New connection pool errors with IllegalStateException on chunked request response setup

### ISSUE THREAD

**Issue by mkrajc** [akka/akka-http#2138]
Upgrading to akka-http 10.1.3 from 10.0.8 IllegalStateException is thrown on chunked request response. It is non deterministic, but happen very often.
```
[ERROR] [08/07/2018 16:26:20.505] [client-akka.actor.default-dispatcher-2] [client/Pool(#1->http://localhost:8080)] [0 (WaitingForEndOfResponseEntity)] Slot execution failed. That's probably a bug. Please file a bug at https://github.com/akka/akka-http/issues. Slot is restarted.
java.lang.IllegalStateException: Cannot [onRequestEntityCompleted] when in state [WaitingForEndOfResponseEntity]
	at akka.http.impl.engine.client.pool.SlotState.illegalState(SlotState.scala:84)
	at akka.http.impl.engine.client.pool.SlotState.onRequestEntityCompleted(SlotState.scala:58)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Event$.$anonfun$onRequestEntityCompleted$1(NewHostConnectionPool.scala:127)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Event$.$anonfun$event0$1(NewHostConnectionPool.scala:143)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.runOneTransition$1(NewHostConnectionPool.scala:209)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.loop$1(NewHostConnectionPool.scala:283)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.updateState(NewHostConnectionPool.scala:292)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.updateState(NewHostConnectionPool.scala:201)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$Slot.onRequestEntityCompleted(NewHostConnectionPool.scala:177)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$pushRequest$3(NewHostConnectionPool.scala:384)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$pushRequest$3$adapted(NewHostConnectionPool.scala:384)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.withSlot(NewHostConnectionPool.scala:373)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$pushRequest$2(NewHostConnectionPool.scala:384)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1$SlotConnection.$anonfun$pushRequest$2$adapted(NewHostConnectionPool.scala:383)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1.$anonfun$safely$2(NewHostConnectionPool.scala:507)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1.$anonfun$safeCallback$1(NewHostConnectionPool.scala:506)
	at akka.http.impl.engine.client.pool.NewHostConnectionPool$HostConnectionPoolStage$$anon$1.$anonfun$safeCallback$1$adapted(NewHostConnectionPool.scala:506)
	at akka.stream.impl.fusing.GraphInterpreter.runAsyncInput(GraphInterpreter.scala:454)
	at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:468)
	at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:563)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:745)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$shortCircuitBatch(ActorGraphInterpreter.scala:735)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:761)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:670)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)
	at akka.actor.ActorCell.invoke(ActorCell.scala:557)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
	at akka.dispatch.Mailbox.run(Mailbox.scala:225)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
Despite the exception is thrown, the response handling for the request is finished ok, but  when getting this error multiple times it will result in pool overflow (default max-open-requests: 32). So it seems like slot is not released.

Here is minimal dummy setup I managed to replicate the issue (similar to our application behaviour). Meanwhile we switched to legacy pool and working without problem.

```
object Server extends Directives {
  def main(args: Array[String]): Unit = {
    implicit val as = ActorSystem("server")
    implicit val am = ActorMaterializer()
    implicit val ec = as.dispatcher

    Http().bindAndHandle(route, "localhost", 8080)
  }

  val maxFrameSize: Int = 20 * 1024

  implicit lazy val jsonStreamingSupport: JsonEntityStreamingSupport =
    EntityStreamingSupport
      .json(maxFrameSize)
      .withParallelMarshalling(parallelism = 2, unordered = false)
      .withFramingRenderer(Flow[ByteString].intersperse(ByteString.empty, ByteString("\n"), ByteString.empty))

  def route: Route = {
    extractRequest { req =>
      entity(asSourceOf[ByteString]) { req =>
        val echoReplace = req.mapAsyncUnordered(4) { r =>
          FastFuture.successful(ByteString(r.utf8String.replace("request", "response")))
        }
        val response = HttpResponse().withEntity(HttpEntity.Chunked.fromData(contentType = `application/json`, echoReplace))
        complete(response)
      }
    }
  }

}

object Client {

  def main(args: Array[String]): Unit = {
    implicit val as = ActorSystem("client")
    implicit val am = ActorMaterializer()
    implicit val ec = as.dispatcher

    val data = Source.fromIterator(() => Iterator.range(0, 1000)).map { i => ByteString("""{"request":""" + i + "}\n") }
    val httpRequest = HttpRequest(POST, uri = "http://localhost:8080/test", entity = HttpEntity(contentType = `application/json`, data))

    val pool = Http().newHostConnectionPool[Int]("localhost", 8080)
    Source.single((httpRequest, 0)).via(pool).flatMapConcat[ByteString, NotUsed] {
      case (Success(resp), r) =>
        resp.entity.dataBytes.mapMaterializedValue(_ => NotUsed).wireTap(s => println(s.utf8String))
      case res =>
        Source.empty[ByteString]
    }
      .runWith(Sink.ignore)
      .andThen {
        case _ => as.terminate()
      }
  }

}

```

**Comment by raboof** [comment]
Hmm, this sounds related to #1746

**Comment by raboof** [comment]
Looking at the code this seems like an oversight: in `WaitingForEndOfResponseEntity` we have a boolean `waitingForEndOfRequestEntity` in which we keep track of whether we still expect the end of the request entity, so it seems when this boolean is on we should accept `onRequestEntityCompleted` (and set this flag to 'false').

**Comment by jrudolph** [comment]
Fixed by #2141.
